feature: SourceCreateRequest validation

MikelAlejoBR opened the pull request:
Validates the "SourceCreateRequest" struct's fields following the guidelines set by the [legacy Ruby code](https://github.com/RedHatInsights/sources-api/blob/d1d7d5dfefab699fe30853f82ec0a3d401ad27dd/app/models/source.rb#L8-L24).


[[RHCLOUD-16353]](https://issues.redhat.com/browse/RHCLOUD-16353)

MikelAlejoBR committed:
feature: SourceCreateRequest validation

MikelAlejoBR committed:
feature: interface to int64 converter

The "toInt64" converter has been extracted to its own function, instead
of having that logic embedded in the "validate" method. This way it can
be reused for the rest of "*Create" methods as well.

MikelAlejoBR committed:
refactor: replace type switch with extracted method

Also removes the "Parsing" related test method, as the parsing is
already being tested on a different file. However, a similar test method
is added to check that the proper error message is returned when that
parsing failed.

MikelAlejoBR committed:
feature: slice search utility functions

As the "SearchString" function from the "sort" package returns the index
in which the target string should be inserted, a little helper code is
needed to actually return a boolean indicating the search result.

Also, the helper "Sort*" function is there to also provide a easy way of
searching a target string in a slice which might be unsorted.

MikelAlejoBR committed:
refactor: move constants to "source" package

MikelAlejoBR committed:
refactor: use "search in slice" instead of long ifs

MikelAlejoBR committed:
refactor: remove redundant test

The test is already being done in the "parser" tests

MikelAlejoBR committed:
feature: enable validation when creating sources

lindgrenj6 commented on model/source_http.go:
```diff
@@ -14,7 +21,48 @@ type SourceCreateRequest struct {
+		req.Uid = &stringId
+	}
+
+	var validWorkflowStatuses = []string{AccountAuth, ManualConfig}
```
We can make this a `var` at the package level - that way it is only alloc'd once rather than every time we're validating something (wish we could make it a const - but we can't do constant slices/arrays :sob:)

Same with the availability status array - but throw that (+ the consts above) in a file called `availability_status.go`, since that will be shared code between basically all of the data models.

MikelAlejoBR replied:
All right. I've left `validWorkflowStatuses` slice on the `source_http.go`, because after doing a quick lookup I haven't found any more usages of `AppCreationWorkflow` on the repository. I'm thinking that maybe I should move the constants of that array to the `source_http.go` file, since I feel it makes more sense to have everything in the same file, instead of having the constants on `source.go` and then the slice in `source_http.go`. What do you think?

lindgrenj6 commented on util/slice_utils.go:
```diff
@@ -0,0 +1,17 @@
+
+// IsStringPresentInSlice returns true if the specified target is present in the given slice. Requires the slice to
+// be ordered in ascending order, due to the underlying requirement with SearchStrings.
+func IsStringPresentInSlice(target string, slice []string) bool {
```
For this function signature - can we reverse the order of the params (and probably rename to the opposite too)? Main reason for asking is thats how `strings.HasPrefix(s, prefix string)` accepts parameters plus how [many other APIs](https://pkg.go.dev/strings@go1.17.1) in Go are structured. It would be good to continue that here.

MikelAlejoBR replied:
I knew this was going to come back at me. I felt SO weird writing that function like that... like after doing everything I was "pff, this feels odd". Should have changed it before.

lindgrenj6 commented on util/slice_utils.go:
```diff
@@ -0,0 +1,17 @@
+func IsStringPresentInSlice(target string, slice []string) bool {
+	index := sort.SearchStrings(slice, target)
+
+	return index < len(slice) && slice[index] == target
```
I'm not a huge fan of the `sort` package specifically for this design (it doesn't return false, it returns an index which we then have to check). 

It does look like it does a binary search which theoretically should be faster: https://cs.opensource.google/go/go/+/refs/tags/go1.17:src/sort/search.go;drc=refs%2Ftags%2Fgo1.17;l=59

But the fact that we have to sort the array before searching negates any benefits imo.

Can we change this to just a for loop looping through the array? That way if the item isn't found we just return false rather than having to check the index and the value at said index.

MikelAlejoBR replied:
Yeah, I was really unsure about using `sort` too because of that "needs the slice to be sorted" requirement... All right, changed the implementation to a simple for loop.

lindgrenj6 commented on util/parser.go:
```diff
@@ -0,0 +1,48 @@
+	"strconv"
+)
+
+// InterfaceToInt64 takes an interface and returns an int64 for any float64, int64 or string received -whether they
+// are pointers or not-.
+func InterfaceToInt64(i interface{}) (int64, error) {
```
This is great, exactly what I was looking for!

MikelAlejoBR replied:
Great :tada:

lindgrenj6 reviewed (commented):
Awesome first pass - this is going to be a great pattern established for the entire app.

Great job on the tests!

MikelAlejoBR committed:
refactor: declare availability status slice at package level

This way we avoid instantiating it every time we have to validate a
new incoming request.

MikelAlejoBR committed:
refactor: declare valid workflow stasues' slice at package level

This way we avoid instantiating it every time it is needed for a validation.

MikelAlejoBR committed:
refactor: rename "search slice in go" function

The arguments are swapped to follow the convention of other packages[1]
and the name of the function is changed accordingly.

https://pkg.go.dev/strings@go1.17.1

MikelAlejoBR commented on model/source_http.go:
```diff
@@ -14,7 +25,46 @@ type SourceCreateRequest struct {
+	}
+
+	// If no valid UUID is provided in the request, we generate one
+	if req.Uid == nil || *req.Uid == "" {
```
Here the code doesn't check that the given UUID is well-formatted UUID\[1\]. I was thinking on adding a regex to check that, what do you think @lindgrenj6 ?

\[1\]: https://en.wikipedia.org/wiki/Universally_unique_identifier#Format

MikelAlejoBR replied:
My bad, the UUID should be automatically generated[1], because the spec doesn't say anything about accepting an `uuid` key in the request body.

\[1\]: https://github.com/RedHatInsights/sources-api/blob/d1d7d5dfefab699fe30853f82ec0a3d401ad27dd/app/models/source.rb#L8

lpichler commented on model/source_http.go:
```diff
@@ -14,7 +25,46 @@ type SourceCreateRequest struct {
+// Validate validates that the required fields of the SourceCreateRequest request hold proper values. In the specific
+// case of the UUID, if an empty or nil one is provided, a new random UUID is generated and appended to the request.
+func (req *SourceCreateRequest) Validate() error {
+	if req.Name == nil || *req.Name == "" {
```
@MikelAlejoBR why it requires both checks ? isn't enough just e.g. `req.Name == nil` ?

MikelAlejoBR replied:
The Rails app actually had an `:allow_blank => false` for the `:name` attribute, so that's why we're performing both checks: [models/source.rb#L23](https://github.com/RedHatInsights/sources-api/blob/5befbc497719db35758488535a44afcb5f82000b/app/models/source.rb#L22)

lpichler commented on model/source_http.go:
```diff
@@ -14,7 +25,46 @@ type SourceCreateRequest struct {
+		req.Uid = &stringId
+	}
+
+	if !util.SliceContainsString(validWorkflowStatuses, req.AppCreationWorkflow) {
```
in ruby version of sources api is also valid when `app_creation_workflow` is not specified - it defaults to `manual_configuration`

MikelAlejoBR replied:
[You are right](https://github.com/RedHatInsights/sources-api/blob/5befbc497719db35758488535a44afcb5f82000b/db/migrate/20210125173413_add_app_creation_workflow_to_sources.rb#L3). Thanks for pointing this out, I'll default the parameter to `manual_configuration` if it comes empty / nil.

MikelAlejoBR committed:
refactor: set "manual_configuration" as a default workflow value

The rails app defaults the value of the "AppCreationWorkflow" to
"manual_configuration" if none is specified[1], so I'm changing the Go
code to match that.

[1]: https://github.com/RedHatInsights/sources-api/blob/5befbc497719db35758488535a44afcb5f82000b/db/migrate/20210125173413_add_app_creation_workflow_to_sources.rb#L3

MikelAlejoBR committed:
fix: inject "TenantID" from the middleware in the DAO

Insertions were being rejected because the required tenant id was
missing. That id is injected in the middleware, and that is why I simply
assign it to the entity to be created.

MikelAlejoBR committed:
refactor: generate a new UUID for every new source created

The UUIDs are not received as part of the request, and therefore that
field will always be nill or empty. That's why it makes sense to
generate a new one every time.

MikelAlejoBR committed:
refactor: move source validation to "service" package

I was getting "circular imports" issues when trying to import the "dao"
package in the "model" package —because the model package then imports
the "dao" package back, thus creating the aforementioned issue". Moving
the validation to a different package will allow us to use both the
"dao"s and the "model"s to properly validate the incoming data.

MikelAlejoBR committed:
feature: check source name's uniqueness in tenant

lpichler commented on dao/source_dao.go:
```diff
@@ -89,3 +90,11 @@ func (s *SourceDaoImpl) Delete(id *int64) error {
+
+func (s *SourceDaoImpl) NameExistsInCurrentTenant(name string) bool {
+	src := &m.Source{Name: name}
+	result := DB.Where("name LIKE ? AND tenant_id = ?", name, s.TenantID).First(src)
```
Can we have "=" instead of LIKE here ?

MikelAlejoBR replied:
Good idea. Done!

MikelAlejoBR committed:
refactor: use "=" instead of "LIKE" operator

lpichler reviewed (approved):
LGTM 👍 thanks!
